### ISSUE TITLE
React Native Guide/Example

### ISSUE THREAD

**Issue by D4nte** [status-im/js-waku#628]
This is a **feature or change request**

## Problem

It is currently not possible to use js-waku in react-native. Probably because of the polyfills.

## Definition of Done

Provide guide and example of using js-waku in React native.

**Comment by jrmeurer** [comment]
+1 this would be awesome

**Comment by D4nte** [comment]
I believe that #337 needs to be done first, especially removal of polyfills.
What kind of incompatibilities are you encountering?

**Comment by jrmeurer** [comment]
```
Unable to resolve module libp2p-gossipsub/src/heartbeat from /Users/.../node_modules/js-waku/build/main/lib/waku_relay/relay_heartbeat.js: libp2p-gossipsub/src/heartbeat could not be found within the project or in these directories:
  node_modules/js-waku/node_modules
  node_modules

If you are sure the module exists, try these steps:
 1. Clear watchman watches: watchman watch-del-all
 2. Delete node_modules and run yarn install
 3. Reset Metro's cache: yarn start --reset-cache
 4. Remove the cache: rm -rf /tmp/metro-*
  25 | Object.defineProperty(exports, "__esModule", { value: true });
  26 | exports.RelayHeartbeat = void 0;
> 27 | const heartbeat_1 = require("libp2p-gossipsub/src/heartbeat");
     |                              ^
  28 | const utils_1 = require("libp2p-gossipsub/src/utils");
  29 | const constants = __importStar(require("./constants"));
  30 | const get_relay_peers_1 = require("./get_relay_peers");

RCTFatal
__28-[RCTCxxBridge handleError:]_block_invoke
_dispatch_call_block_and_release
_dispatch_client_callout
_dispatch_main_queue_drain
...
```

**Comment by D4nte** [comment]
> ```
> Unable to resolve module libp2p-gossipsub/src/heartbeat from /Users/.../node_modules/js-waku/build/main/lib/waku_relay/relay_heartbeat.js: libp2p-gossipsub/src/heartbeat could not be found within the project or in these directories:
> ```

Yes,  a dependency broke their API on patch version. Should be fixed with `js-waku@0.20.0`. Let me know if not.

**Comment by jrmeurer** [comment]
Now:  
`ERROR  Error: Requiring module "node_modules/js-waku/build/main/index.js", which threw an exception: TypeError: Conversion from 'BigInt' to 'number' is not allowed.
`

Same error we had with @noble/ed25519 library. We were not able to resolve this with either polyfills or babel configs.

**Comment by D4nte** [comment]
> Now: `ERROR Error: Requiring module "node_modules/js-waku/build/main/index.js", which threw an exception: TypeError: Conversion from 'BigInt' to 'number' is not allowed. `
> 
> Same error we had with @noble/ed25519 library. We were not able to resolve this with either polyfills or babel configs.

Thanks @jrmeurer. I believe @jemboh has reached the same error.
He is off this week (back Thursday).

I wonder if adding a shim.js to js-waku could be a solution, as described here: https://github.com/facebook/react-native/issues/28492#issuecomment-824698934

It means changing js-waku of course. Is that something you already tried? FYI @jemboh, I think we should try this next, WDYT?

**Comment by jrmeurer** [comment]
Shim in js-waku could be a good idea. I haven't tried this yet - we have a massive deadline in 1 month, so a bit distracted. 
Would love to launch with js-waku, I imagine it will be a group effort 🙏

**Comment by jemboh** [comment]
@jrmeurer I've been working on this and have come across the same issue with the `BigInt` error. I've written a rough guide in the readme. https://github.com/status-im/rn-waku-relay-poc

I'm going to look in to a shim in `js-waku` itself as well. 

Any extra things you might have seen along the way, please feel free to raise a PR or an issue in that repo so we can keep each other updated on our progress.

Thanks for all your help and work so far! 🤝

**Comment by D4nte** [comment]
It may make more sense to use a native module based on go-waku. See https://discord.com/channels/864066763682218004/865466694554484738/971941000333393960

**Comment by D4nte** [comment]
For now we are focusing effort on publishing a native module using go-waku.

We are still open to making js-waku work in react-native but not clear how this can be achieved:

https://discord.com/channels/864066763682218004/865466694554484738/977027312782573578

**Comment by D4nte** [comment]
Waku is now available for react-native platforms via a go-waku native module: https://discord.com/channels/864066763682218004/957040003593154571/981324886679113808

`yarn add @waku/react-native`

Join Discord and give us some feedback.

Here are my latest thoughts on js-waku in react native: https://discord.com/channels/864066763682218004/865466694554484738/977027312782573578

>  wonder if we could have a way to make @noble/secp256k1 and its react native for optional dependencies and simply switch the library used depending on availability.
> js-waku is currently using webpack but I am open to change it.

I am currently changing to rollup that supports the `browser` field in `package.json`. Will need to look up if it also support the `react-native` field.

**Comment by jrmeurer** [comment]
> I am currently changing to rollup that supports the `browser` field in `package.json`. Will need to look up if it also support the `react-native` field.

Incredible work, friends. 💪
react-native field is supported in package.json.

**Comment by D4nte** [comment]
> Incredible work, friends. muscle react-native field is supported in package.json.

Supported by whom? Webpack support does not seem to be out of the box.

**Comment by D4nte** [comment]
#680 should hopefully help with that.

**Comment by fryorcraken** [comment]
Let's park it as https://github.com/waku-org/waku-react-native is available.

Should now work as the issues are fixed:
- no more polyfills
- Latest React Native supports BigInt AFAIK
- nobles library have replaced BigInt's `n` and `**` with `BigInt` and `.pow`

Can be reviewed if a platform wants to use the same SDK for Browser and React Native.

**Comment by weboko** [comment]
Bringing it back to discuss potential work on the topic as it is asked by people who want to use `js-waku`.

cc @waku-org/js-waku-developers @fryorcraken

**Comment by maschad** [comment]
Not sure if this is the best place to comment but I am currently leading an effort to do this for [js-libp2p](https://github.com/libp2p/js-libp2p/issues/2136) which would be required at a minimum to utilize it in js-waku, I can't speak to what further lift would be required after that as I am not so familiar with js-waku's codebase.

**Comment by fryorcraken** [comment]
cc @b4s36t4

**Comment by weboko** [comment]
> Bringing it back to discuss potential work on the topic as it is asked by people who want to use `js-waku`.
> 
> cc @waku-org/js-waku-developers @fryorcraken
> 
> If we don't plan to enable it and repo https://github.com/waku-org/waku-react-native is not working then we need to deprecate it and update README + docs to prevent people from spending time on unnecessary things

Decoupling important part of this task into https://github.com/waku-org/js-waku/issues/1743

The example part is probably better to shape into `template` and to do in scope of https://github.com/waku-org/js-waku-examples/issues/271

**Comment by danisharora099** [comment]
> Some progress on react-native compatibility for js-libp2p - [libp2p/js-libp2p#2136 (comment)](https://github.com/libp2p/js-libp2p/issues/2136#issuecomment-1849999462)

super cool! 

--- 
imo we should prioritize the react-native SDK effort with js-waku cc @waku-org/js-waku-developers @chair28980

**Comment by weboko** [comment]
Adding a reference to @maschad example https://github.com/ipfs-shipyard/js-libp2p-react-native

**Comment by weboko** [comment]
What is done - pollyfills are provided - https://github.com/waku-org/js-waku/pull/1915
Leftover at this stage is to provide an example https://github.com/waku-org/lab.waku.org/pull/36

Moving to TODO for now.

**Comment by fryorcraken** [comment]
Work in this should be paused for now.

**Comment by weboko** [comment]
> Work in this should be paused for now.

I moved it into TODO because of that. Do you think we move to Icebox?
